### PR TITLE
add stream agg executor.

### PR DESCRIPTION
**Pull request by hanfei1991**
I will split #1686 into multiple small prs. This pr only implents stream agg executor.
@zimulala @coocood @shenli @tiancaiamao @XuHuaiyu  PTAL

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1111,6 +1111,124 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+			break
+		}
+	}
+	if !e.hasData && len(e.GroupByItems) > 0 {
+		return nil, nil
+	}
+	return retRow, nil
+}
+
+func (e *StreamAggExec) updateGroupKey(row *Row) (bool, error) {
```
Comment (shenli) [inline executor/executor.go]: Add comment for the first return value.

Comment (shenli) [inline executor/executor.go]: meetNewGroup() maybe better

File: executor/executor.go [hunk 2]
```diff
@@ -1111,6 +1111,124 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+	if len(e.curGroupKey) == 0 {
+		matched, firstGroup = false, true
+	}
+	for i, item := range e.GroupByItems {
+		v, err := item.Eval(row.Data, e.ctx)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if matched {
+			c, err := v.CompareDatum(e.curGroupKey[i])
```
Comment (shenli) [inline executor/executor.go]: This is used to avoid calling codec.EncodeValue() for every row?

File: expression/aggregation.go [hunk 3]
```diff
@@ -213,18 +325,63 @@ func (cf *countFunction) GetGroupResult(groupKey []byte) (d types.Datum) {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ctx.Count += count.GetInt64()
+	return nil
+}
+
 // Update implements AggregationFunction interface.
 func (af *avgFunction) Update(row []types.Datum, groupKey []byte, ctx context.Context) error {
+	if af.mode == FinalMode && af.name == ast.AggFuncAvg {
```
Comment (shenli) [inline expression/aggregation.go]: Could avgFunction.name be any other value except ast.AggFuncAvg?

File: executor/executor.go (hunk 2 above)
Comment (hanfei1991) [inline executor/executor.go]: yes

Comment (shenli) [inline executor/executor.go]: Please add comment for this trick.

File: expression/aggregation.go (hunk 3 above)
Comment (hanfei1991) [inline expression/aggregation.go]: no, It's my fault

**Comment by shenli** [comment]
@coocood @tiancaiamao @zimulala PTAL

File: executor/executor.go [hunk 4]
```diff
@@ -1111,6 +1111,125 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+			matched = c == 0
+		}
+		e.tmpGroupKeyBuffer = append(e.tmpGroupKeyBuffer, v)
+	}
+	if matched {
+		return false, nil
+	}
+	e.curGroupKey = e.tmpGroupKeyBuffer
+	var err error
+	e.curGroupEncodedKey, err = codec.EncodeValue(e.curGroupEncodedKey[0:0:cap(e.curGroupEncodedKey)], e.tmpGroupKeyBuffer...)
```
Comment (coocood) [inline executor/executor.go]: `e.tmpGroupKeyBuffer` -> `e.curGroupKey` makes more sense.

File: executor/executor.go [hunk 5]
```diff
@@ -1111,6 +1111,125 @@ func (e *AggregationExec) innerNext() (ret bool, err error) {
...
+	schema             expression.Schema
+	ResultFields       []*ast.ResultField
+	executed           bool
+	hasData            bool
+	ctx                context.Context
+	AggFuncs           []expression.AggregationFunction
+	GroupByItems       []expression.Expression
+	curGroupEncodedKey []byte
+	curGroupKey        []types.Datum
+	tmpGroupKeyBuffer  []types.Datum
```
Comment (coocood) [inline executor/executor.go]: `tmpGroupKeyBuffer` -> `tmpGroupKey`, as `buffer` is usually used in bytes type.

File: expression/aggregation.go [hunk 6]
```diff
@@ -76,11 +85,23 @@ func NewAggFunction(funcType string, funcArgs []Expression, distinct bool) Aggre
 
 type aggCtxMapper map[string]*ast.AggEvaluateContext
 
+// AggFunctionMode stands for the aggregation function's mode.
+type AggFunctionMode int
+
+const (
+	// CompleteMode function accepts origin data.
+	CompleteMode AggFunctionMode = iota
```
Comment (coocood) [inline expression/aggregation.go]: `CompleteMode` is a little misleading, looks like a mode that is special.
How about rename it to `OriginMode` according to the comment?

Comment (coocood) [inline expression/aggregation.go]: Or if there is only two modes, and we will never add new mode in the future, we can just set `FinalMode` as a bool property in aggFunction.

File: expression/aggregation.go [hunk 7]
```diff
@@ -280,6 +452,41 @@ func (cf *concatFunction) Update(row []types.Datum, groupKey []byte, ectx contex
 	return nil
 }
 
+func (cf *concatFunction) StreamUpdate(row []types.Datum, ectx context.Context) error {
+	ctx := cf.getStreamedContext()
+	vals := make([]interface{}, 0, len(cf.Args))
```
Comment (coocood) [inline expression/aggregation.go]: Why not define `vals` as `[]types.Datum`, and use `ToString`?
I'm afraid for some types like `mysql.Time` or `mysql.MyDeciaml`, the result of `fmt.Sprintf("%v", val` would be incorrect.
Probably this is not covered in test.

**Comment by tiancaiamao** [comment]
LGTM

File: expression/aggregation.go (hunk 6 above)
Comment (hanfei1991) [inline expression/aggregation.go]: The name of "Complete Mode" is referenced by Hive.

File: expression/aggregation.go (hunk 7 above)
Comment (hanfei1991) [inline expression/aggregation.go]: This is ported from "Update" function. We will consider this case in next pr.

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

File: plan/column_pruning.go [hunk 8]
```diff
@@ -127,14 +127,16 @@ func (p *Aggregation) PruneColumnsAndResolveIndices(parentUsedCols []*expression
 		return nil, errors.Trace(err)
 	}
 	for _, aggrFunc := range p.AggFuncs {
-		for i, arg := range aggrFunc.GetArgs() {
+		var newArgs []expression.Expression
```
Comment (zimulala) [inline plan/column_pruning.go]: It can make use `len(aggrFunc.GetArgs())`

Comment (hanfei1991) [inline plan/column_pruning.go]: Will fix in #1686

**Pull request merged**